drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
-            ),
-            (RotComponent::Hubris, SlotId::B) => (
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-                __IMAGE_B_BASE.as_ptr() as u32..__IMAGE_B_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::A) => (
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
-            (RotComponent::Stage0, SlotId::B) => (
-                __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
-                __IMAGE_STAGE0_BASE.as_ptr() as u32
-                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
-            ),
+/// Accessor keeps the implementation details of ImageAccess private
+enum Accessor<'a> {
+    // Flash driver, flash device range
+    Flash {
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+    Ram {
+        buffer: &'a [u8],
+        span: FlashRange,
+    },
+    // Hybrid is used for later implementation of rollback protection.
+    // The buffer is used in place of the beginning of the flash range.
+    _Hybrid {
+        buffer: &'a [u8],
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        span: FlashRange,
+    },
+}
+
+impl Accessor<'_> {
+    fn exec(&self) -> &Range<u32> {
+        match self {
+            Accessor::Flash { flash: _, span }
+            | Accessor::Ram { buffer: _, span }
+            | Accessor::_Hybrid {
+                buffer: _,
+                flash: _,
+                span,
+            } => &span.exec,
+        }
+    }
+}
+
+/// The update_server needs to deal with images that are:
+///   - Entirely in flash
+///   - Header block in RAM with the remainder unavailable
+///   - Header block in RAM with the remainder in flash.
+///   - Entire image in RAM (cached Stage0)
+/// Calls to methods use image relative addresses.
+/// ImageAccess::Flash{_, span} gives the physical flash addresses
+/// being accessed.
+///
+/// Any address here is the "storage" address which for stage0next
+/// and any image held temporarily in RAM is different than the
+/// execution address.
+pub struct ImageAccess<'a> {
+    accessor: Accessor<'a>,
+}
+
+impl ImageAccess<'_> {
+    pub fn new_flash<'a>(
+        flash: &'a drv_lpc55_flash::Flash<'a>,
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'a> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Flash { flash, span },
+        }
+    }
+
+    pub fn new_ram(
+        buffer: &[u8],
+        component: RotComponent,
+        slot: SlotId,
+    ) -> ImageAccess<'_> {
+        let span = flash_range(component, slot);
+        ImageAccess {
+            accessor: Accessor::Ram { buffer, span },
```

Comment:
(Here is a place where the length-matches invariant is definitely not maintained, fwiw.)

Reply:
Unless all of a slot is in use, it is expected that buffer.len() would always be less than the length of one of the spans.

Perhaps I should change `span` to  `slot_range` (since `Slot` is already in use).
Suggestions welcome.

Span is a FlashRange which describes both the stored and the at_runtime address ranges of the slot that the image can occupy.
`buffer` here is a portion of the image that currently resides in RAM. The rest of the image may not be present.
In the case of a Accessor::Ram variant, the entire image resides in ram. It's destination slot is described by `span.stored` and its runtime slot is described by `span.at_runtime` (new names).